modules/hivemq-edge-module-opcua/src/main/java/com/hivemq/edge/adapters/opcua/OpcUaProtocolAdapter.java

```diff
@@ -107,6 +109,38 @@ public OpcUaProtocolAdapter(
         this.opcUaClientConnection = new AtomicReference<>();
     }
 
+    /**
+     * Calculates backoff delay based on the number of consecutive retry attempts.
+     * Parses the comma-separated retryIntervalMs string and returns the appropriate delay.
+     * If attemptCount exceeds the number of configured delays, returns the last configured delay.
+     *
+     * @param retryIntervalMs comma-separated string of backoff delays in milliseconds
+     * @param attemptCount the number of consecutive retry attempts (1-indexed)
+     * @return the backoff delay in milliseconds
+     * @throws NumberFormatException when the format is incorrect
+     */
+    public static long calculateBackoffDelayMs(final @NotNull String retryIntervalMs, final int attemptCount) {
+        final String[] delayStrings = retryIntervalMs.split(",");
```

Comment:
the String could be parsed in the start method, just once, instead of on every retry.